middlewares/cors.go

```diff
@@ -0,0 +1,179 @@
+package middlewares
+
+import (
+	"github.com/savsgio/atreugo/v10"
+	"github.com/savsgio/go-logger"
+
+	"strconv"
+	"strings"
+)
+
+type CorsOptions struct {
+	AllowedOrigins   []string
+	AllowedHeaders   []string
+	AllowMaxAge      int
+	AllowedMethods   []string
+	ExposedHeaders   []string
+	AllowCredentials bool
+}
+
+type CorsHandler struct {
+	maxAge            int
+	allowCredentials  bool
+	allowedOriginsAll bool
+	allowedHeadersAll bool
+	allowedOrigins    []string
+	allowedHeaders    []string
+	allowedMethods    []string
+	exposedHeaders    []string
+}
+
+var defaultOptions = &CorsOptions{
+	AllowedOrigins: []string{"*"},
+	AllowedMethods: []string{"GET", "POST"},
+	AllowedHeaders: []string{"Origin", "Accept", "Content-Type"},
+}
+
+func DefaultCors() *CorsHandler {
+	return NewCorsMiddleware(*defaultOptions)
+}
+
+func NewCorsMiddleware(options CorsOptions) *CorsHandler {
+	cors := &CorsHandler{
+		allowedOrigins:   options.AllowedOrigins,
+		allowedHeaders:   options.AllowedHeaders,
+		allowCredentials: options.AllowCredentials,
+		allowedMethods:   options.AllowedMethods,
+		exposedHeaders:   options.ExposedHeaders,
+		maxAge:           options.AllowMaxAge,
+	}
+
+	if len(cors.allowedOrigins) == 0 {
+		cors.allowedOrigins = defaultOptions.AllowedOrigins
+		cors.allowedOriginsAll = true
+	} else {
+		for _, v := range options.AllowedOrigins {
+			if v == "*" {
+				cors.allowedOrigins = defaultOptions.AllowedOrigins
+				cors.allowedOriginsAll = true
+				break
+			}
+		}
+	}
+
+	if len(cors.allowedHeaders) == 0 {
+		cors.allowedHeaders = defaultOptions.AllowedHeaders
+		cors.allowedHeadersAll = true
+	} else {
+		for _, v := range options.AllowedHeaders {
+			if v == "*" {
+				cors.allowedHeadersAll = true
+				break
+			}
+		}
+	}
+
+	if len(cors.allowedMethods) == 0 {
+		cors.allowedMethods = defaultOptions.AllowedMethods
+	}
+
+	return cors
+}
+
+func (c *CorsHandler) CorsMiddleware(ctx *atreugo.RequestCtx) {
+	c.handlePreflight(ctx)
+}
+
+func (c *CorsHandler) handlePreflight(ctx *atreugo.RequestCtx) {
+	originHeader := string(ctx.Request.Header.Peek("Origin"))
+
+	if len(originHeader) == 0 || !c.isAllowedOrigin(originHeader) {
+		logger.Debug("Origin ", originHeader, " is not in", c.allowedOrigins)
+		return
+	}
+
+	method := string(ctx.Request.Header.Method())
+	if !c.isAllowedMethod(method) {
+		logger.Debug("Method ", method, " is not in", c.allowedMethods)
+		return
+	}
+
+	var headers []string
+	if len(ctx.Request.Header.Peek("Access-Control-Request-Headers")) > 0 {
+		headers = strings.Split(string(ctx.Request.Header.Peek("Access-Control-Request-Headers")), ",")
+	}
+
+	if !c.areHeadersAllowed(headers) {
+		logger.Debug("Headers ", headers, " is not in", c.allowedHeaders)
+		return
+	}
+
+	ctx.Response.Header.Set("Access-Control-Allow-Origin", originHeader)
+	ctx.Response.Header.Set("Access-Control-Allow-Methods", method)
```

Comment:
PUT and DELETE methods fails because when a preflight requests is sent, the `Access-Control-Allow-Methods` value is `OPTIONS` and you must set the all allowed methods, not the current method, which is `OPTIONS` on preflight.

GET and POST are allowed by default by the browers so it's not mandatory to put in `Access-Control-Allow-Methods`, but another methods must be in  `Access-Control-Allow-Methods` header if you want to works.